core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java

```diff
@@ -96,20 +104,34 @@ private void calculateData(IItemCollection input) {
 			IAggregator<?, ?> aggregator = Aggregators.distinct(query.getGroupBy());
 			Set<?> aggregate = input.getAggregate((IAggregator<Set<?>, ?>) aggregator);
 			if (aggregate != null) {
-				for (Object o : aggregate) {
-					IItemCollection rowCollection = input.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
-					Object[] row = newRow();
-					int column = 0;
-					for (; column < attributes.size(); column++) {
-						// Optimization - it is too expensive to do aggregation for these. You simply
-						// get first non-null matching attribute - we're only using this for the group by today.
-						row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+				int SUM_NUM_THREADS = Runtime.getRuntime().availableProcessors();
+				ExecutorService exec = Executors.newFixedThreadPool(SUM_NUM_THREADS);
```

Comment:
I wonder if a `ForkJoinPool` might be better suited here (maybe with `Executors.newWorkStealingPool`), in which case the code won't need to look for the number of avaialble processors.
Going furhter it might be even useful to consider `Stream::parallelStream`?